Add basic translation capabilities to MerC

SilverMight opened the pull request:
This PR adds basic translation capabilities to MerC.

`emit_translations.py` handles the generation of translated sources files by looking at the input source directory, taking a Maki results file and calling into `analyze_transformations.py` to generate transformations, and modifies the file as needed.

`run_maki_on_compile_commands.py`  is script to iteratively run Maki over a `compile_commands.json` file and output results into a single results file that can be used with `emit_translations.py`.

Transformation capabilities are currently limited and only work under certain conditions (i.e one definition per macro, in a global scope). Further PRs will expand scope of translatable macros.

SilverMight committed:
Add function to generate syntactic translations of macros

SilverMight committed:
Script for running Maki on multiple commands and getting output files

SilverMight committed:
Modularize analyze_transformations, add emit_translations to emit translations for directory of maki output files

SilverMight committed:
Multithread and fix paths

SilverMight committed:
Add logic to read in Maki out and translate files

SilverMight committed:
Handle ICE invocations, ignore duplicate defintions

We need to translation object-like macros used where ICEs are required
to anonymous enums, as const variables are not ICEs.

In addition, currently ignore macros that have multiple definitions.

SilverMight committed:
Refactor I/O of Maki analysis to one, deduped file

Previously we made multiple Maki analysis files to match the source
directory structure, however this is unnecessary. Combine everything
into one, deduplicated file (most important for compiler builtins).

In addition, emit_translations now takes in a single Maki file and no
longer a directory of results (flag: "maki_results_dir" -> "maki_results_path")

Smaller changes in run_maki_on_compile_commands.py are a new flag for
the number of threads to us and a fix to make sure we're in the correct
directory while running each compile command.

SilverMight committed:
Add flags to Maki to ignore builtin macros

SilverMight committed:
Properly handle translations of object-like macros
1) Add new IsICERepresentableByInt32 property from Maki to Invocation class.
2) Translation fixes:
    * If any invocations require an ICE and are not representable by an int, skip translation.
    * If all invocations that require ICE are representable by int, translate to enum.
    * Otherwise translate to const static variable of same type.

SilverMight committed:
Implement global scope check from Maki
1) Add IsDefinedAtGlobalScope property to Macro class.
2) Add check to return false is we are not at global scope in
   interface_equivalent predicate.
3) Simplify macro construction

SilverMight committed:
Cleanup unused variables, imports, functions
1) Remove only function and InvocationPredicate as they're unused
2) Optimize imports
3) Fix whitespace

SilverMight committed:
Remove main from analyze_transforms & reformatting

SilverMight committed:
Add logging and fix arg parse bug
Replace print statements with logger statements, which we have more
flexibility over output levels with. Users can change level to info with
new -v/--verbose flag.

SilverMight committed:
Refactor running Maki to use ProcessPoolExecutor
1) Use a ProcessPoolExecutor to run Maki.
2) Have run_maki_on_compile_command return json instead of putting it
   into a queue, and remove the queue.
   * No need to order anything and better expresses intent and trims
     down function signature
3) Trim some unnecessary variables from function signature

SilverMight committed:
Clean up arg names and add help descriptions

SilverMight committed:
Fix types and adjust exception handling
1) Remove case where we exited after capturing any exception. Printing
   the traceback would be more helpful.
2) Change return types for generate_macro_translations to match that
   there could be no translation (str | None).
3) Change MacroMap to be a dict instead of a default dict (we only
   iterate over what's placed in it)
4) Change incorrect json type to dict[str, Any]
5) Improve logging in run maki on compile commands and call .decode() on
   process stderr/stdout.

SilverMight committed:
Add missing import and remove unused

Bottle-Water committed:
bug fixes on fstring formatting and clang call

PappasBrent commented on analyze_transformations.py:
```diff
@@ -46,36 +32,80 @@ def easy_to_transform_definition(m: Macro,
+        # Static to avoid breaking the one definition rule
+        if macro.IsFunctionLike:
+            # return only if not void
+            returnStatement = "return" if not invocation.TypeSignature.startswith("void") else ""
```
This would trigger for `void *` return types too and cause us not to emit `return` statements for macros which expand to `void *` types. I think this would be a broken translation.

SilverMight replied:
Good catch. Fixed in latest commit

PappasBrent commented on analyze_transformations.py:
```diff
@@ -46,36 +32,80 @@ def easy_to_transform_definition(m: Macro,
+            # All invocations where an ICE is required must be representable by type int 
+            # to be translatable to an enum
+            can_translate_to_enum = all(
+                [i.IsICERepresentableByInt32 for i in invocations if i.IsInvokedWhereICERequired])
```
You know I realize now that the standard only says an `int`, however on some old platforms an `int` is only 16 bytes... maybe we should restrict ourselves further? It *probably* doesn't matter but we are dealing with old code that is meant to be portable so maybe it does.

SilverMight replied:
Definitely worth considering, the standard only guarantees at least 16-bits for an integer. Maybe it's worth having it configurable?

[A good article](https://thephd.dev/c23-is-coming-here-is-what-is-on-the-menu#n3029---improved-normal-enumerations) that talks about the use of this in some codebaes (and how they fixed it in C23).

PappasBrent replied:
Oh yeah I've seen that one before! I don't think I finished reading it though but I should.

PappasBrent commented on analyze_transformations.py:
```diff
@@ -46,36 +32,80 @@ def easy_to_transform_definition(m: Macro,
+                [i.IsICERepresentableByInt32 for i in invocations if i.IsInvokedWhereICERequired])
+
+            # If no invocations require ICE, just make it a static const variable
+            not_invoked_where_ICE_required = all([not i.IsInvokedWhereICERequired for i in invocations])
```
We'll probably have to update this to use the new property I added to Maki, `IsInvokedWhereConstantExpressionRequired`, to handle macros defined to non-integral values.

SilverMight replied:
Will handle in a separate PR

PappasBrent commented on analyze_transformations.py:
```diff
@@ -46,36 +32,80 @@ def easy_to_transform_definition(m: Macro,
+    with open(results_file) as fp:
         entries = json.load(fp)
 
+        # Keep only one macro definition. if there's more than one, keep none
```
May want to elaborate on *why* we keep only keep macros that are defined exactly once (I presume because of multiple translation issues you had mentioned in a prior meeting)

PappasBrent commented on analyze_transformations.py:
```diff
@@ -46,36 +32,80 @@ def easy_to_transform_definition(m: Macro,
         entries = json.load(fp)
 
+        # Keep only one macro definition. if there's more than one, keep none
+        for obj in entries:
+            if obj["Kind"] == "Definition":
+                if obj["Name"] in unique_names:
+                    unique_names[obj["Name"]] = None
+                else:
+                    unique_names[obj["Name"]] = obj
```
This seems like a good place for a `Counter()`, but this approach should work as well.

SilverMight replied:
Actually this doesn't work right now and I didn't commit my changes to fix it, which did use Counter. Will commit them soon.

PappasBrent commented on macros.py:
```diff
@@ -264,11 +265,11 @@ def SatisfiesALanguageSpecificProperty(self) -> bool:
 @dataclass
 class PreprocessorData:
-    mm: MacroMap = field(default_factory=lambda: defaultdict(set))
+    mm: MacroMap = field(default_factory=dict)
```
Why switch to a `dict` here?

SilverMight replied:
For our usage (generate_macro_translations), we are just iterating over the keys and values present in the dictionary after MacroMap is constructed. I don't think we do any arbitrary accesses so we shouldn't access a key that doesn't exist.

This also caused a lot of type-checking errors in get_interface_equivalent_preprocessordata as we were not constructing defaultdicts (src_pd, tlna_src_pd, ie_pd)

PappasBrent commented on emit_translations.py:
```diff
@@ -0,0 +1,87 @@
+
+def translate_src_files(src_dir: str, out_dir: str, translations: dict[Macro, str]) -> None:
+    # dict of src files to their contents in lines
+    src_file_contents: dict[str, list[str]] = {}
```
Nice optimization to avoid reading files multiple times.

SilverMight committed:
Fix filtering of macros with multiple definitions
Logic was incorrect before and did not properly filter macros that show
up more than once.
Also abstract this into a function and add some comments explaining why
we do this.

SilverMight committed:
Fix return statement check for function-like macro
Old check did not consider that we still need to return for `void*`. Added
regex to make sure we match `void` but not `void *`.
